engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/IntermediateCatchEventProcessor.java

```diff
@@ -41,36 +49,141 @@ public Class<ExecutableCatchEventElement> getType() {
-            })
-        .ifRightOrLeft(
-            completed -> stateTransitionBehavior.takeOutgoingSequenceFlows(element, completed),
-            failure -> incidentBehavior.createIncident(failure, completing));
+    eventBehaviorOf(element).onComplete(element, completing);
   }
 
   @Override
   public void onTerminate(
       final ExecutableCatchEventElement element, final BpmnElementContext terminating) {
+    eventBehaviorOf(element).onTerminate(element, terminating);
+
+    // common behavior for all intermediate catch events
     eventSubscriptionBehavior.unsubscribeFromEvents(terminating);
     incidentBehavior.resolveIncidents(terminating);
 
     final var terminated = stateTransitionBehavior.transitionToTerminated(terminating);
     stateTransitionBehavior.onElementTerminated(element, terminated);
   }
+
+  private IntermediateCatchEventBehavior eventBehaviorOf(
+      final ExecutableCatchEventElement element) {
+    return catchEventBehaviors.stream()
+        .filter(behavior -> behavior.isSuitableForEvent(element))
+        .findFirst()
+        .orElseThrow(
+            () ->
+                new UnsupportedOperationException(
+                    "This kind of intermediate catch event is not supported."));
+  }
+
+  private interface IntermediateCatchEventBehavior {
+
+    boolean isSuitableForEvent(final ExecutableCatchEventElement element);
+
+    void onActivate(final ExecutableCatchEventElement element, final BpmnElementContext activating);
+
+    default void onComplete(
+        final ExecutableCatchEventElement element, final BpmnElementContext completing) {}
+
+    default void onTerminate(
+        final ExecutableCatchEventElement element, final BpmnElementContext terminating) {}
+  }
+
+  private class NoneIntermediateCatchEventBehavior implements IntermediateCatchEventBehavior {
+
+    @Override
+    public boolean isSuitableForEvent(final ExecutableCatchEventElement element) {
+      return element.isNone();
+    }
+
+    @Override
+    public void onActivate(
+        final ExecutableCatchEventElement element, final BpmnElementContext activating) {
+      final var activated = stateTransitionBehavior.transitionToActivated(activating);
+      stateTransitionBehavior.completeElement(activated);
+    }
+
+    @Override
+    public void onComplete(
+        final ExecutableCatchEventElement element, final BpmnElementContext completing) {
+      variableMappingBehavior
+          .applyOutputMappings(completing, element)
+          .flatMap(ok -> stateTransitionBehavior.transitionToCompleted(element, completing))
+          .ifRightOrLeft(
+              completed -> stateTransitionBehavior.takeOutgoingSequenceFlows(element, completed),
+              failure -> incidentBehavior.createIncident(failure, completing));
+    }
+  }
+
+  private abstract class AbstractIntermediateCatchEventBehavior
+      implements IntermediateCatchEventBehavior {
```

Comment:
:wrench: We should favor delegation over inheritance. 

Instead of having an abstract behavior class, we could have a default behavior, like the none behavior. The default behavior could be used for all event types, except the link event. 

The logic of `onComplete()` is the same for all events and could be shared.